helpers/misc.go

```diff
@@ -18,26 +19,61 @@ func BoolPtr(b bool) *bool {
 	return &b
 }
 
-// Retry will retry a function until it succeeds or the timeout is reached. timeout == 2^attempt * delay.
-func Retry(fn func() error, retries int, delay time.Duration, logger func(format string, args ...any)) error {
+// RetryWithContext retries a function until it succeeds, the timeout is reached, or the context is done.
+// The delay between attempts increases exponentially as (2^(attempt-1)) * delay.
+// For example, with a delay of one second and three attempts, the timing would be:
+// - First attempt: immediate
+// - Second attempt: after one second
+// - Third attempt: after two seconds
+func RetryWithContext(ctx context.Context, fn func() error, attempts int, delay time.Duration, logger func(format string, args ...any)) error {
 	var err error
-	for r := 0; r < retries; r++ {
```

Comment:
Something I forgot to comment on, currently this code will not perform the action at all `if attempts < 1 (0, -1, etc...)`. We should prob either A: `error`, or B: set `attempts = 1`, not sure if there is an option C.

not sure if we should throw anything into the `logger` to signal back that no retries happened

```go
	t.Run("NoRetry", func(t *testing.T) {
		count := 0
		fn := func() error {
			count++
			return nil
		}
		logger := func(_ string, _ ...any) {}

		err := Retry(fn, -1, 0, logger)
		require.ErrorIs(t, err, nil)
		require.Equal(t, 1, count)
	})

Reply:
IMO returning an error like "no retries happened" is enough without logging

Reply:
Implemented